source/common/router/router.cc

```diff
@@ -636,8 +659,14 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
   // Hang onto the modify_headers function for later use in handling upstream responses.
   modify_headers_ = modify_headers;
 
-  UpstreamRequestPtr upstream_request =
-      std::make_unique<UpstreamRequest>(*this, std::make_unique<HttpConnPool>(*http_pool));
+  UpstreamRequestPtr upstream_request;
+  if (!should_tcp_proxy) {
```

Comment:
Not sure if I get it: When shouldProxy return false, it could be either 
1) non-CONNECT request, or
2) CONNECT but vHost doest support CONNECT
IMHO 1) should use http conn pool, however, why should 2) use http connpool?

Reply:
That's a really good question.

I planned on calling it out in the next PR (and adding changes to the router if folks didn't like my plan A) but happy to chat now.

Docs for the connect feature are up at my WIP PR here:
https://github.com/envoyproxy/envoy/compare/master...alyssawilk:shinkansen_next

but in summary I thought it would be useful for Envoy to be able to proxy CONNECT requests untouched (forward the connect headers and payload) as well as terminate CONNECT requests.  For folks with a multi-proxy set up they may want to forward encapsulated TCP over multiple hops before they "downgrade" it to TCP, especially if they're using HTTP/2 multiplexing and can avoid all the single use TCP connections

I have the proxying-CONNECT (vs terminating CONNECT behavior) as default because honestly I find CONNECT terrifying - all it takes is one config slip and you're forwarding unsanitized HTTP headers including admin commands at your upstreams.  Eek!  I'd rather folks who screw up their config end up with unexpected CONNECT requests at their upstream rather than unexpected decapuslated unsanitized HTTP at their upstreams.  

If we think this is a bad call since CONNECT is by default supposed to be hop by hop, we could fail CONNECT requests over non-CONNECT-enabled-routes by default, unless there's a route-specific override, but that gets more complicated and IMO is less safe.  So let's discuss :-)

@PiotrSikora  @mattklein123

Reply:
See my comment above about the API/config. Let's chat there first?